File: algorithms/algorithms-cplusplus/leetcode/Question_0055_Jump_Game.cc
#include <vector>

using namespace std;

class Solution {
 public:
  bool canJump(vector<int>& nums) {
    if (nums.empty()) {
      return false;
    }
    int m = 0;
    for (int i = 0; i < nums.size(); i++) {
      m = max(m, i + nums[i]);
      if (m < i + 1) {
        return false;
      }
      if (m >= nums.size() - 1) {
        return true;
      }
    }
    return false;
  }
};

File: algorithms/algorithms-python/leetcode/Question_0055_Jump_Game.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        if not nums:
            return False
        if len(nums) == 1:
            return True
        
        m = 0
        n = len(nums) - 1
        for i in range(n):
            m = max(m, i + nums[i])
            if m < i + 1:
                return False
            if m >= n:
                return True
        return False
